Approved/dyn-type.md

```diff
@@ -0,0 +1,196 @@
+A Dyn Type For Runtime Values
+=====
+
+## Proposal
+
+This proposal introduces a `dyn` type to reflect values that are only known at runtime.
+
+## Description
+
+The `dyn` type doesn't require any user annotation as it is inferred by the compiler. The `dyn` type is used with global callables. `dyn` values come from intrinsic operations. Functions can't introduce `dyn` values, but can pass them. The Result type is always a `dyn`. A `dyn` Result can be converted to a `dyn` Bool.  The `dyn` static analysis replaces mutable variable and result analysis.
+
+## Use Cases
+
+Below use cases explain how a developer interacts with the `dyn` type and describe the functionalities of the new type.
+
+### Allow to pass the values created at runtime to other callables
+
+```qsharp
+operation Reset(q : Qubit, m : Result) : Unit {
+    if m == One {
+        X(q);
+    }
+}
+
+operation MeasureAndReset(q : Qubit) : Unit {
+    let res = M(q); // create a dynamic value (res)
+    Reset(q, res); // pass the dynamic value to another operation
+}
+```
+
+Without the proposed modification, the implementation of `Reset` would need to be manually inlined into `MeasureAndReset`.
+
+### Allow callables to return the values created at runtime
+
+```qsharp
+operation MeasureAsInt(q : Qubit) : Int {
+    return M(q) == One ? 1 | 0;
+}
+
+operation CountOnes(qs : Qubit[]) : Int {
+    mutable count = 0;
+    for q in qs {
+        set count += MeasureAsInt(q);
+    }
+    return count;
+}
+```
+
+Without the proposed modification, dynamic values can only be returned by the entry point, but not by other callables.
+
+Without introducing dynamic types, callables defined in libraries could only be used for computing compile time known values, but not for manipulating runtime values; Ultimately, application code then would need to be mostly contained in the entry point callable and code cannot be structured into several subroutines.
+
+### Dynamic values require runtime support
+
+The example generates an error when calling `Foo` with a dynamic value as argument `i` unless the target supports arrays.
+
+The developer can create values at runtime by measuring a qubit and then pass these values to other callables that process it further and return it.
+
+> Define the array index as a callable input parameter
+
+```qsharp
+operation Foo(b : Bool, i : Int) : Int {
+    return [1, 2, 3][i]; // This needs runtime support if i is dynamic
+}
+
+operation Bar() : Int {
+// get a dynamic value (usually by measuring a qubit)
+    use q = Qubit();
+    H(q);
+    let val = M(q) == Zero ? 0 : 1; // val is of type dyn Int
+
+    let res = Foo(true, val); // requires that the runtime supports arrays
+    return res; // res is of type dyn Int
+}
+```
+
+> Define the lifting operator for the array index
+
+```qsharp
+lift_i Foo : (Bool, dyn Int) => dyn Int
```

Comment:
Is this literal Q# code? If so, that would seem to imply that `dyn` is part of the syntax?